source/common/router/shadow_writer_impl.cc

```diff
@@ -6,17 +6,25 @@
 #include "common/common/assert.h"
 #include "common/http/headers.h"
 
+#include "absl/strings/match.h"
+#include "absl/strings/str_join.h"
+
 namespace Envoy {
 namespace Router {
 
 void ShadowWriterImpl::shadow(const std::string& cluster, Http::MessagePtr&& request,
                               std::chrono::milliseconds timeout) {
+  ASSERT(request->headers().Host()->value().c_str() != "");
+
   // Switch authority to add a shadow postfix. This allows upstream logging to make a more sense.
-  // TODO PERF: Avoid copy.
-  std::string host = request->headers().Host()->value().c_str();
-  ASSERT(!host.empty());
-  host += "-shadow";
-  request->headers().Host()->value(host);
+  if (absl::StrContains(request->headers().Host()->value().c_str(), ":")) {
+    auto parts = StringUtil::splitToken(request->headers().Host()->value().c_str(), ":");
+    ASSERT(parts.size() == 2);
+    request->headers().Host()->value(absl::StrJoin(parts, "-shadow:"));
+  } else {
+    request->headers().Host()->value(absl::StrJoin(
```

Comment:
absl::StrCat() would work here instead of requiring forward_as_tuple()

Reply:
Have it fixed in 7d7e613. While I'm still not sure about the assertions.